docs/spec.md

```diff
@@ -3102,7 +3125,8 @@ separate outputs to improve clarity
 
 Fills an `output` tensor with values in increasing order starting from zero
 along the `iota_dimension` dimension. More formally, `output[result_index] =
-constant(result_index[iota_dimension], element_type(output))`.
+constant(result_index[iota_dimension], is_quantized(output) ?
+        storage_type(output) : element_type(output))`.
```

Comment:
I think the specification here is different from what people (or quantizer) expect when we quantize a float tensor into a quantized tensor.

For example, a floating point iota:

```
    %a = "stablehlo.iota"() {iota_dimension = 1} : () -> tensor<2x8xf32>
```

Returns:

```
[[0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0],
 [0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0]]
```

If we quantize this float tensor with scale = 2, zp = 0 (with round to nearest even rounding), the storage-typed tensor should be:

```
[[0, 0, 1, 2, 2, 2, 3, 4],
 [0, 0, 1, 2, 2, 2, 3, 4]]
```

But if I interpret this line correctly, this is specifying:

```
[[0, 1, 2, 3, 4, 5, 6, 7],
 [0, 1, 2, 3, 4, 5, 6, 7]]
```

If this storage-typed tensor is dequantized, the value will be very different from the float tensor.

Reply:
I completely agree with you that the quantized iota is just referring to the storage type and how that is going to be interpreted in the floating-point world depends on the choice (user-driven) of the quantization parameters. For example, with zp=0, scale=1, one can expect to get the desired correspondence in floating-point world.  

Having said that, IMO, we can keep the current specification as in the specification mainly because we made it clear in the spec that the stored value are of type `storage_type` . What do you think?

Reply:
Update the ops `collective_permute`, `convolution`, `dot_general`, and `iota`. Now we have,

* `collective_permute`
    * Updated text: `broadcast_in_dim(constant(zero_point(result), element_type(result)), [], type(result))`
* `convolution`: 
    * Same as before: `padding_value = constant(is_quantized_tensor(lhs) ? zero_point(lhs) : 0, element_type(lhs))`.
* `dot_general`:
   * Updated text: `init_values=[constant(0, storage_type(result)]`
   * Rationale: The parent reduce op, which the `init_values` corresponds to, is doing integer computation. 
* `iota`:
  * Update test as suggested above.   

PTAL.

Reply:
dot_general & iota look fine. Now that I looked at this again, I'm realizing that `reshaped_lhs_slice - zero_point(reshaped_lhs_slice)` doesn't typecheck, but we can leave resolving this to some other venue - this seems out of scope for this PR.

collective_permute - How would this work for non-quantized types?  `zero_point(result)` is going to return `None` in that case, which is probably not what we want.

convolution - This notation no longer works now that we have specified that QuantizedTensorConstant "has_syntax(quantized_tensor_literal, storage_type(quantized_tensor_type))". This is why I proposed writing things up explicitly for iota and why I chose to use `quantize` in that proposal.

Reply:
>> collective_permute - How would this work for non-quantized types? zero_point(result) is going to return None in that case, which is probably not what we want.

That's my bad. fixed it. 

>> convolution - This notation no longer works

Just want to make sure I get it right. `constant(zero_point(lhs), element_type(lhs))` does not work because the type of `zero_point(lhs)` is not `element_type(lhs)`. To ensure the type parity your proposal is to explicitly express the quantize operation. 

With the above understanding, I have changed  
-* `padding_value = constant(is_quantized(lhs) ? zero_point(lhs) : 0, element_type(lhs))`.

to

+* `padding_value = constant(is_quantized(lhs) ? quantize(0, element_type(lhs)) : 0, element_type(lhs))`.

Please let me know if I am missing something.

Reply:
LGTM.